File: src/gaet/remote.py
"""Management of remote cloud database configuration (git remote parity)."""

import argparse
import json
import sys
from typing import Dict, Any, Optional

from .core import (
    load_env,
    set_env_key,
    get_env_str,
    parse_remote_url,
    find_pg_tools,
    pg_env,
    cleanup_pg_env,
    run_cmd,
    echo,
    die,
    box_title,
    box_section,
    status_ok,
    status_arrow,
    status_warn,
    set_output_modes,
    C, G, Y, R, D, B, NC,
    DEF_REMOTE_SSLMODE,
)
from .registry import command


def cmd_remote(args: argparse.Namespace) -> None:
    """Manage remote cloud DB settings (show / set-url / remove)."""
    subaction = getattr(args, "remote_action", None) or "show"
    url_arg = getattr(args, "url", None)
    want_json = getattr(args, "json", False)

    if want_json:
        set_output_modes(quiet=True, plain=True)

    env = load_env()

    if subaction == "set-url":
        if not url_arg:
            die("Usage: gaet remote set-url <postgresql://user:pass@host:port/db>")
        parsed = parse_remote_url(url_arg)
        if not parsed:
            die("Invalid remote URL format. Expected: postgresql://user:pass@host:port/dbname")
        set_env_key("GAET_REMOTE_URL", url_arg)
        status_ok(f"GAET_REMOTE_URL updated successfully ({parsed['user']}@{parsed['host']}:{parsed['port']}/{parsed['db']})")
        return

    if subaction in ("remove", "unset", "rm"):
        set_env_key("GAET_REMOTE_URL", "")
        status_ok("GAET_REMOTE_URL removed successfully from .env")
        return

    # Default / 'show' action
    remote_url = get_env_str(env, "GAET_REMOTE_URL") or get_env_str(env, "GAET_SUPABASE_URL") or ""
    parsed = parse_remote_url(remote_url)

    if want_json:
        result = {
            "command": "remote",
            "configured": bool(parsed),
            "remote": parsed if parsed else None,
        }
        print(json.dumps(result, indent=2))
        return

    box_title("gaet remote")
    if not parsed:
        echo(f"  {Y}No Remote Cloud DB configured yet.{NC}")
        echo(f"  Usage: {C}gaet remote set-url postgresql://user:pass@host:port/db{NC}")
        return

    masked_pass = "*****" if parsed["pass"] else ""
    masked_url = f"postgresql://{parsed['user']}:{masked_pass}@{parsed['host']}:{parsed['port']}/{parsed['db']}"

    box_section("Remote Configuration")
    status_arrow(f"Host:     {parsed['host']}")
    status_arrow(f"Port:     {parsed['port']}")
    status_arrow(f"User:     {parsed['user']}")
    status_arrow(f"Database: {parsed['db']}")
    status_arrow(f"URL:      {masked_url}")

    # Connectivity Check
    echo()
    echo(f"  {C}☁️{NC}   Testing remote cloud connection... ", end="", flush=True)
    tools = find_pg_tools(env)
    psql = tools.get("psql", "")
    if psql:
        ssl = get_env_str(env, "GAET_REMOTE_SSLMODE", DEF_REMOTE_SSLMODE)
        env_remote = pg_env(parsed["user"], parsed["pass"], ssl)
        try:
            out, _, rc = run_cmd(
                [psql, "-w", "-h", parsed["host"], "-p", parsed["port"],
                 "-U", parsed["user"], "-d", parsed["db"], "-tAc", "SELECT 1;"],
                env=env_remote, timeout=5,
            )
        finally:
            cleanup_pg_env(env_remote)
        if rc == 0 and out.strip() == "1":
            echo(f"{G}OK{NC}")
        else:
            echo(f"{R}FAIL{NC}")
            status_warn("Cannot connect to remote cloud DB. Check your connection URL or network.")
    else:
        echo(f"{Y}SKIPPED (psql not found){NC}")
    echo()


def _build_remote_parser(subparsers, common):
    p = subparsers.add_parser("remote", help="Manage remote cloud database configuration", parents=[common])
    p.add_argument("remote_action", nargs="?", default="show", choices=["show", "set-url", "remove", "unset", "rm"],
                   help="Action: show (default), set-url, remove")
    p.add_argument("url", nargs="?", default=None, help="Connection URL (for set-url)")
    p.add_argument("--json", action="store_true", help="Output JSON result")
    return p


command("remote", "Manage remote cloud database configuration", build=_build_remote_parser)(cmd_remote)

File: pkg/remote/remote.go
// Package remote implements `gaet remote` command.
package remote

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/ghanirahmans/gaet/pkg/core"
)

// RunRemote implements `gaet remote [action] [url]`.
func RunRemote(action, urlArg string, jsonOut bool) error {
	if jsonOut {
		core.Quiet = true
		core.Plain = true
	}

	env, err := core.LoadEnv(core.EnvFile())
	if err != nil {
		return err
	}

	switch strings.ToLower(action) {
	case "set-url":
		if urlArg == "" {
			return core.Die("Usage: gaet remote set-url <postgresql://user:pass@host:port/db>", core.ExitUsage)
		}
		if _, err := core.ParseRemoteURL(urlArg); err != nil {
			return core.Die("Invalid remote URL format. Expected: postgresql://user:pass@host:port/dbname", core.ExitConfig)
		}
		if err := core.SetEnvKey(core.EnvFile(), "GAET_REMOTE_URL", urlArg); err != nil {
			return err
		}
		core.WriteJSONLog(core.LogEntry{
			Level:    "INFO",
			Category: "REMOTE",
			Action:   "REMOTE_SET",
			Status:   "SUCCESS",
			Message:  "GAET_REMOTE_URL updated successfully",
		})
		core.StatusOK("GAET_REMOTE_URL updated successfully")
		return nil

	case "remove", "unset", "rm":
		if err := core.SetEnvKey(core.EnvFile(), "GAET_REMOTE_URL", ""); err != nil {
			return err
		}
		core.WriteJSONLog(core.LogEntry{
			Level:    "INFO",
			Category: "REMOTE",
			Action:   "REMOTE_REMOVE",
			Status:   "SUCCESS",
			Message:  "GAET_REMOTE_URL removed from .env",
		})
		core.StatusOK("GAET_REMOTE_URL removed from .env")
		return nil

	default: // "show"
		remoteURL := core.GetEnvStr(env, "GAET_REMOTE_URL", "")
		if remoteURL == "" {
			remoteURL = core.GetEnvStr(env, "GAET_SUPABASE_URL", "")
		}
		parsed, parseErr := core.ParseRemoteURL(remoteURL)

		if jsonOut {
			result := map[string]any{"command": "remote", "configured": parseErr == nil, "remote": nil}
			if parseErr == nil {
				result["remote"] = map[string]any{
					"host": parsed.Host, "port": parsed.Port,
					"user": parsed.User, "db": parsed.DB,
				}
			}
			enc := json.NewEncoder(os.Stdout)
			enc.SetIndent("", "  ")
			return enc.Encode(result)
		}

		core.BoxTitle("gaet remote")
		if parseErr != nil {
			core.StatusWarn("No Remote Cloud DB configured yet.")
			core.Echo(fmt.Sprintf("  Usage: %sgaet remote set-url postgresql://user:pass@host:port/db%s",
				core.ColorCyan, core.ColorReset))
			return nil
		}

		maskedURL := core.MaskURLPassword(remoteURL)
		core.BoxSection("Remote Configuration")
		core.StatusArrow(fmt.Sprintf("Host:     %s", parsed.Host))
		core.StatusArrow(fmt.Sprintf("Port:     %s", parsed.Port))
		core.StatusArrow(fmt.Sprintf("User:     %s", parsed.User))
		core.StatusArrow(fmt.Sprintf("Database: %s", parsed.DB))
		core.StatusArrow(fmt.Sprintf("URL:      %s", maskedURL))

		fmt.Println()
		core.Echo(fmt.Sprintf("  %s[INFO]%s  Testing remote cloud connection...",
			core.ColorBCyan, core.ColorReset))
		tools := core.FindPGTools(env)
		if tools.Psql != "" {
			ssl := core.GetEnvStr(env, "GAET_REMOTE_SSLMODE", core.DefRemoteSSLMode)
			envCloud := core.PGEnv(parsed.User, parsed.Password, ssl)
			out, _, rc := core.RunCmdSimple(tools.Psql,
				[]string{"-w", "-h", parsed.Host, "-p", parsed.Port,
					"-U", parsed.User, "-d", parsed.DB, "-tAc", "SELECT 1;"},
				envCloud, 5*time.Second)
			if rc == 0 && strings.TrimSpace(out) == "1" {
				core.StatusOK("Cloud connection OK")
			} else {
				core.StatusFail("Cannot connect to cloud database")
				core.StatusWarn("Check GAET_REMOTE_URL and cloud database status")
			}
		} else {
			core.StatusWarn("psql not found — connection test skipped")
		}
		fmt.Println()
		return nil
	}
}
